### ISSUE TITLE
/usr/local/bin/gitx crash after upgrading to 1.0

### ISSUE THREAD

**Issue by fbzhong** [gitx/gitx#397]
I just upgrade gitx to 1.0 and found gitx will crash at startup.

Version 1.0 (1.0.2925 [arm])

```bash
❯ gitx .
2023-04-11 15:31:11.548 gitx[29480:791546] warning: failed to get scripting definition from /Applications/GitX.app/Contents/Frameworks/Sparkle.framework/Versions/B/Updater.app; it may not be scriptable.
2023-04-11 15:31:11.548 gitx[29480:791546] -[SBApplication open:withOptions:]: unrecognized selector sent to instance 0x6000003493e0
2023-04-11 15:31:11.548 gitx[29480:791546] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[SBApplication open:withOptions:]: unrecognized selector sent to instance 0x6000003493e0'
*** First throw call stack:
(
        0   CoreFoundation                      0x000000019c89319c __exceptionPreprocess + 176
        1   libobjc.A.dylib                     0x000000019c3b24d4 objc_exception_throw + 60
        2   CoreFoundation                      0x000000019c93a178 -[NSObject(NSObject) __retain_OA] + 0
        3   CoreFoundation                      0x000000019c7fb150 ___forwarding___ + 1600
        4   CoreFoundation                      0x000000019c7faa50 _CF_forwarding_prep_0 + 96
        5   gitx                                0x000000010499960c gitx + 5644
        6   gitx                                0x000000010499a4f0 gitx + 9456
        7   dyld                                0x000000019c3e3f28 start + 2236
)
libc++abi: terminating due to uncaught exception of type NSException
zsh: abort      gitx .
```

**Comment by hannesa2** [comment]
From which version do you upgraded ?
It's somehow unrelated: I deinstalled gitx and installed from brew and it works properly

**Comment by hannesa2** [comment]
1. I de-installed my brew gitx by delete gitx from Applications
2. Installed https://github.com/gitx/gitx/releases/download/0.23/GitX-arm64.dmg
3. Run menu "Enable Terminal usage"
4. Install https://github.com/gitx/gitx/releases/download/1.0/GitX-arm64.dmg by drag gitx to Application <img width="499" alt="image" src="https://user-images.githubusercontent.com/3314607/231128047-72837815-ec5f-4c21-a75a-84a55814a80c.png">
6. Run menu "Enable Terminal usage"

--> all works fine. It looks like the issue is located on your side

**Comment by hopkinschris** [comment]
For whatever it's worth, I encountered the exact same issue as @fbzhong when "upgrading" to `1.0` from `0.21`. Reverted back to `0.21` for the meantime...

**Comment by hannesa2** [comment]
As far as I know, only https://github.com/gitx/gitx/releases/tag/0.22 introduced an issue with notarization.
Anything else is unknown to me. Who understands the call stack. I don't understand

**Comment by eirc** [comment]
Had the same issue, deleted my previous gitx terminal binary with `sudo rm /usr/local/bin/gitx`, run gitx, clicked enable terminal usage and it's fine now.

**Comment by fbzhong** [comment]
> Had the same issue, deleted my previous gitx terminal binary with `sudo rm /usr/local/bin/gitx`, run gitx, clicked enable terminal usage and it's fine now.

I tried but still crashed.

**Comment by fbzhong** [comment]
@fredngo You are genius! I download from release and copy to /Applications again, then follow your instructions, it works now!

**Comment by chief1983** [comment]
So, I was able to reproduce the fix above, but why did I not get prompted when running the app via Cmd+Space?  If it had not let me run it at all I would have gone to Finder and Right-Click -> Open'd on the freshly downloaded app.  But since it ran fine I didn't think there was anything else to be done confirmation-wise.  I never use Launchpad for anything except monitoring Xcode upgrades as those can take hours.  It's some significant UX friction for a declared 1.0 release.

**Comment by hmeine** [comment]
Not sure this issue should be closed, though - as long as people still run into it there will be duplicates. And AFAICS we still do not fully understand how this happens.

**Comment by hannesa2** [comment]
> Not sure this issue should be closed, though - as long as people still run into it there will be duplicates. And AFAICS we still do not fully understand how this happens.

I completer agree to this. If someone understands what's going wrong, please tell it. To understand is the first step to solve.

**Comment by fredngo** [comment]
I wonder if there's a way to somehow automate the execution of an `xattr` command to get rid of the `com.apple.quarantine` extended attribute that Mac places on the app when it's downloaded.

**Comment by fredngo** [comment]
@hmeine Yep, that's my hypothesis. I think `/Applications/GitX.app` has the quarantine flag on it, which is set automatically by MacOS when it's downloaded. I don't know how other apps deal with this, I'm not at all a desktop app developer.

**Comment by abasiri** [comment]
I had the same issue. The following steps worked for me:

1) `rm -rf ~/Library/Saved Application State/net.phere.GitX.savedState`
2) `rm ~/Library/Preferences/net.phere.GitX.plist`
3) `rm -rf ~/Library/HTTPStorages/net.phere.GitX`
4) `rm -rf ~/Library/Caches/net.phere.GitX`
5) This step might not be needed: `mkdir <new-dir-name>` and ran `git init` in that directory, then invoked `gitx` in this fresh directory.

After i went back to my regular repo and ran gitx, everything worked.

**Comment by hannesa2** [comment]
Should we close it, or is there more to do ?

**Comment by SwissalpS** [comment]
I'd only close it if there is a commit that fixes it.

**Comment by radiospiel** [comment]
#346 looks like a dupe

**Comment by chief1983** [comment]
I just did an install on a fresh Silicon Macbook Pro and didn't run into a problem.  I thought I used the same steps to try to produce it but somehow avoided it this time, or it's even harder to trigger than I thought.  Doesn't matter to me now if it's closed.

**Comment by hmeine** [comment]
Yes, I just updated from 1.1 to 1.4 in order to test this – the usual DnD installation from the (arm64) dmg image leads to a crash when running `gitx` in the terminal. Running GitX via spotlight works. Re-installing the CLI does not fix the problem, nor did removing quarantine flags after the crash happened once. I had to replace GitX.app from the dmg again and *first* start the bundle before using the CLI.

I wonder if the problem could be also caught in the app;
```
2026-03-27 18:31:40.767 gitx[87215:19714251] warning: failed to get scripting definition from /Applications/GitX.app/Contents/Frameworks/Sparkle.framework/Versions/B/Updater.app; it may not be scriptable.
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[SBApplication open:withOptions:]: unrecognized selector sent to instance 0xaf701c0c0'
```
does look like the app should be able to handle the missing scripting definition (and at least give a more informative error message).

**Comment by chief1983** [comment]
I also just ran into it again.  Was about to report a new issue and went through my notification and stumbled across this thread.  Claude had this to say about a possible workaround, since the root cause is that the app is quarantined until explicitly ran from Finder, even though Spotlight launching works just fine due to the app being notarized:

> Q: Yes, I just stumbled upon 397 in my github notifications while you were looking into this, and forgot this is an issue I'd even interacted with before.  Oops.  I had the question there at the time though - why doesn't Spotlight launching generate the prompt?  It literally lets the app run and everything, even with the quarantine.  That seems like a flaw, but I could have sworn other Spotlight launched apps didn't work that way?
> 
> That's a good question and it is a bit counterintuitive. The distinction is subtle but intentional on Apple's part:
> 
> Spotlight launch goes through NSWorkspace/Launch Services, which does check Gatekeeper — but for a notarized app (which GitX is), it silently passes. Notarization satisfies the "is this safe to run?" check, so no prompt is shown and the app launches fine. But the quarantine xattr stays on the files.
> 
> Finder right-click > Open (or double-click triggering the prompt) does something extra — it actually removes the quarantine attribute from the bundle after you confirm. This is the "user has explicitly reviewed and accepted this app" gesture that Apple treats as stronger than just "Launch Services verified the notarization."
> 
> The difference matters because ScriptingBridge doesn't care about notarization — it's hitting a lower-level file access check where the quarantine xattr on the embedded .sdef (or on the Sparkle sub-bundle) causes the read to fail.
> 
> So the flow is:
> 
> Spotlight launch: notarization check passes silently, app runs, quarantine xattr persists on disk
> Finder open with prompt: user confirms, quarantine xattr removed, ScriptingBridge can now read the sdef
> You're right that it feels like a flaw — the app is trusted enough to execute but not trusted enough for another process to read its resources. It's arguably a gap in Apple's quarantine model where the "trust" granted by notarization verification doesn't fully propagate to all subsystems that check quarantine flags.
> 
> Other apps you've launched via Spotlight probably didn't have this issue because they either (a) were installed via a .pkg installer which strips quarantine, (b) were installed via Homebrew which runs xattr -cr, or (c) don't rely on ScriptingBridge for their CLI component.
> 
> Q: Can an app tell the user that it is still quarantined when ran from the command line, before it generates that error?
> 
> Yes, and there are a couple of approaches — from simple to proactive:
> 
> Simplest: catch the crash with @try/@catch
> 
> The CLI could wrap the ScriptingBridge calls and give a useful message instead of an abort. In [gitx.m:119-126](vscode-webview://1m7nvk30e5nemaoeq0r1gre0d0ca9ocg30hip6elkgfhmk00172k/Classes/gitx.m#L119-L126):
> 
> 
> ```
> @try {
>     [gitXApp open:repositoryURL withOptions:arguments];
> } @catch (NSException *e) {
>     fprintf(stderr, "Failed to communicate with GitX.app.\n"
>         "This usually means the app is still quarantined.\n"
>         "Open GitX.app from Finder first, then try again.\n");
>     exit(1);
> }
> ```
> That's defensive but reactive — it doesn't detect the problem until it blows up.
> 
> Better: check quarantine proactively
> 
> The CLI could check for the xattr before even attempting the ScriptingBridge call:
> 
> 
> ```
> #include <sys/xattr.h>
> 
> BOOL appIsQuarantined(void) {
>     NSString *appPath = [[NSWorkspace sharedWorkspace]
>         URLForApplicationWithBundleIdentifier:kGitXBundleIdentifier].path;
>     if (!appPath) return NO;
>     char buf[1];
>     return getxattr(appPath.fileSystemRepresentation,
>                     "com.apple.quarantine", buf, sizeof(buf), 0, 0) > 0
>         || errno == ERANGE;  // attribute exists but buffer too small
> }
> ```
> Then before creating the SBApplication, warn and exit cleanly.
> 
> Best: both
> 
> Check proactively with a clear message, and keep the @try/@catch as a safety net for any other ScriptingBridge failures nobody has thought of yet.
> 
> Honestly even just the @try/@catch alone would be a huge improvement — the current experience is a raw Objective-C exception stack trace that means nothing to most users. That would be a straightforward PR if you wanted to contribute it.

I'm also no Mac dev so if that helps someone else with more competence than I have make a PR, or be inspired to find another workaround, that'd be neat.  In the meantime I'll clear the Quarantine again like I did the last time I updated it :P
